src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/ZipVerifiers.java

```diff
@@ -0,0 +1,125 @@
+
+    public static final String DOCUMENTS_ZIP = "documents.zip";
+    public static final String SIGNATURE_SIG = "signature.sig";
+
+    private ZipVerifiers() {
+    }
+
+    public static Function<ZipStreamWithSignature, ZipInputStream> getPreprocessor(
+        String signatureAlgorithm
+    ) {
+        if ("sha256withrsa".equalsIgnoreCase(signatureAlgorithm)) {
+            return ZipVerifiers::sha256WithRsaVerification;
+        } else if ("none".equalsIgnoreCase(signatureAlgorithm)) {
+            return ZipVerifiers::noOpVerification;
+        }
+        throw new RuntimeException("Undefined signature verification algorithm");
+    }
+
+    static ZipInputStream noOpVerification(ZipStreamWithSignature zipWithSignature) {
+        return zipWithSignature.zipInputStream;
+    }
+
+    static ZipInputStream sha256WithRsaVerification(ZipStreamWithSignature zipWithSignature) {
+        Map<String, byte[]> zipEntries = extractZipEntries(zipWithSignature.zipInputStream);
+        if (!verifyFileNames(zipEntries)) {
+            throw new RuntimeException(
+                "Zip entries do not match expected file names. "
+                + "Actual names = " + zipEntries.keySet()
+            );
+        }
+        if (!verifySignature(zipWithSignature.publicKeyBase64, zipEntries)) {
+            throw new RuntimeException("Zip signature failed verification");
+        }
+        return zipWithSignature.zipInputStream;
+    }
+
+    private static Map<String, byte[]> extractZipEntries(ZipInputStream zis) {
+        Map<String, byte[]> zipEntries = new HashMap<>();
+        ZipEntry zipEntry;
+        try {
+            while ((zipEntry = zis.getNextEntry()) != null) {
+                zipEntries.put(zipEntry.getName(), toByteArray(zis));
+            }
+        } catch (IOException ioe) {
+            throw new RuntimeException(ioe);
+        }
+
+        return zipEntries;
+    }
+
+
+    static boolean verifyFileNames(Map<String, byte[]> entries) {
+        return
+            entries.size() == 2
+            &&
+            entries.keySet().stream()
+            .filter(n -> DOCUMENTS_ZIP.equalsIgnoreCase(n) || SIGNATURE_SIG.equalsIgnoreCase(n))
+            .count() == 2;
+    }
+
+    static boolean verifySignature(String publicKeyBase64, Map<String, byte[]> entries) {
+        return verifySignature(
+            publicKeyBase64,
+            entries.get(DOCUMENTS_ZIP),
+            entries.get(SIGNATURE_SIG)
+        );
+    }
+
+    public static boolean verifySignature(String publicKeyBase64, byte[] data, byte[] signed) {
+        PublicKey publicKey = decodePublicKey(publicKeyBase64);
+        try {
+            Signature signature = Signature.getInstance("SHA256withRSA");
+            signature.initVerify(publicKey);
+            signature.update(data);
+            return signature.verify(signed);
+        } catch (NoSuchAlgorithmException | InvalidKeyException | SignatureException e) {
+            throw new RuntimeException(e);
```

Comment:
Better to throw as custom user defined exception here which will include zip file name which failed verification.

Reply:
Please do read the comments on the PR and the ticket before commenting.